File: packages/orchestrator/src/agents/generic-plan.ts
import { existsSync, readdirSync, readFileSync, statSync } from "fs";
import { join } from "path";
import { defaultGenerate, defaultLocate } from "../skills/base.js";
import type { ClarifiedRequest, FileStep, RepoContext, Skill, SkillPlan } from "../types.js";
import type { LLMClient } from "../llm/doubao.js";

const MAX_GENERIC_FILES = 6;

function listDir(root: string, relative: string, depth: number): string[] {
  const dir = join(root, relative);
  if (!existsSync(dir) || depth < 0) return [];

  const out: string[] = [];
  for (const entry of readdirSync(dir).sort()) {
    if (entry === "node_modules" || entry === ".git" || entry === "dist" || entry === "build") continue;
    const rel = relative ? `${relative}/${entry}` : entry;
    const full = join(root, rel);
    const stat = statSync(full);
    out.push(stat.isDirectory() ? `${rel}/` : rel);
    if (stat.isDirectory()) out.push(...listDir(root, rel, depth - 1));
  }
  return out;
}

function safeRead(root: string, relative: string): string {
  const file = join(root, relative);
  if (!existsSync(file)) return "(文件不存在)";
  return readFileSync(file, "utf8").split("\n").slice(0, 180).join("\n");
}

function listFiles(root: string, relative: string): string[] {
  const dir = join(root, relative);
  if (!existsSync(dir)) return [];
  return readdirSync(dir)
    .filter((name) => statSync(join(dir, name)).isFile())
    .sort()
    .map((name) => `${relative}/${name}`);
}

export function getRepoOverview(ctx: RepoContext): string {
  const root = ctx.repoPath;
  return [
    `仓库路径：${root}`,
    `当前分支：${ctx.branch}`,
    "--- 目录树（最多 2 层）---",
    listDir(root, "", 2).slice(0, 220).join("\n"),
    "--- 前端入口文件 ---",
    `--- frontend/src/App.jsx ---\n${safeRead(root, "frontend/src/App.jsx")}`,
    `--- frontend/src/main.jsx ---\n${safeRead(root, "frontend/src/main.jsx")}`,
    "--- 前端页面文件 ---",
    listFiles(root, "frontend/src/routes").join("\n"),
    listFiles(root, "frontend/src/routes/Profile").join("\n"),
    "--- 后端路由注册 ---",
    safeRead(root, "backend/index.js"),
    "--- 后端路由文件 ---",
    listFiles(root, "backend/routes").join("\n"),
    "--- 后端模型文件 ---",
    listFiles(root, "backend/models").join("\n"),
  ].join("\n");
}

function parsePlan(text: string): FileStep[] {
  const cleaned = text.replace(/^```[a-z]*\n?/m, "").replace(/```$/m, "").trim();
  const data = JSON.parse(cleaned) as { files?: Array<Partial<FileStep>> };
  if (!Array.isArray(data.files)) throw new Error("genericPlan 没有返回 files 数组");

  return data.files.slice(0, MAX_GENERIC_FILES).map((f, i) => {
    const path = String(f.path ?? "").trim();
    const mode = f.mode === "create" ? "create" : f.mode === "modify" ? "modify" : null;
    const instruction = String(f.instruction ?? "").trim();
    if (!path) throw new Error(`genericPlan.files[${i}] 缺少 path`);
    if (!mode) throw new Error(`genericPlan.files[${i}] mode 必须是 modify 或 create`);
    if (!instruction) throw new Error(`genericPlan.files[${i}] 缺少 instruction`);
    return { path, mode, instruction };
  });
}

export async function genericPlan(
  req: ClarifiedRequest,
  ctx: RepoContext,
  llm: LLMClient,
): Promise<SkillPlan> {
  const repoOverview = getRepoOverview(ctx);
  const result = await llm.chat([
    {
      role: "system",
      content: "你是资深代码规划助手。只输出严格 JSON，不要输出 markdown 或解释。",
    },
    {
      role: "user",
      content: `分析下面的产品需求，并提出需要修改或新建的文件清单。

需求摘要：${req.summary}
字段名：${req.fieldName}
字段类型：${req.fieldType}
展示位置：${req.displayLocation}
业务规则：${req.businessRule}

仓库概览：
${repoOverview}

输出 JSON，格式必须是：
{"files":[{"path":"相对路径","mode":"modify|create","instruction":"具体实现说明"}]}

要求：
- path 必须相对 Conduit 仓库根目录。
- mode="modify" 时，path 必须是真实存在的文件，或能从仓库概览中明确判断为已存在文件。
- mode="create" 时，path 必须放在符合现有约定的位置。
- 不要超过 6 个文件；典型需求控制在 2-4 个文件。
- instruction 要写清楚本文件具体改什么，不能只写“修改这个文件”。`,
    },
  ], { temperature: 0.1, maxTokens: 1600 }, { agent: "plan:generic" });

  return { skillName: "generic", files: parsePlan(result.text) };
}

export const genericSkill: Skill = {
  name: "generic",
  description: "未命中注册 skill 时使用的通用推理路径",
  possibleAspects: [],
  aspectQuestionTemplate: {},
  match: () => 0,
  async plan() {
    throw new Error("genericSkill.plan 需要 LLM，请直接调用 genericPlan(req, ctx, llm)");
  },
  locate: defaultLocate,
  generate: defaultGenerate,
};

File: packages/orchestrator/orchestrator/agents/generic_plan.py
from __future__ import annotations

import json
import re
from pathlib import Path

from orchestrator.skills.base import Skill
from orchestrator.types import FileStep, RepoContext

MAX_GENERIC_FILES = 12


def _list_dir(root: Path, relative: str, depth: int) -> list[str]:
    path = root / relative
    if not path.exists() or depth < 0:
        return []
    out: list[str] = []
    for child in sorted(path.iterdir(), key=lambda p: p.name):
        if child.name in {"node_modules", ".git", "dist", "build"}:
            continue
        rel = str(child.relative_to(root)).replace("\\", "/")
        out.append(rel + "/" if child.is_dir() else rel)
        if child.is_dir():
            out.extend(_list_dir(root, rel, depth - 1))
    return out


def _safe_read(root: Path, relative: str) -> str:
    path = root / relative
    if not path.exists():
        return "(文件不存在)"
    return "\n".join(path.read_text(encoding="utf-8", errors="ignore").splitlines()[:180])


def _list_files(root: Path, relative: str) -> list[str]:
    path = root / relative
    if not path.exists():
        return []
    return sorted(str(p.relative_to(root)).replace("\\", "/") for p in path.iterdir() if p.is_file())


def _existing_db_tables(root: Path) -> list[str]:
    """从既有 migration 的 createTable 调用里提取真实表名，供新建/修改 migration 照抄（避免大小写/复数臆造）。"""
    names: set[str] = set()
    mig_dir = root / "backend" / "migrations"
    if not mig_dir.exists():
        return []
    for f in sorted(mig_dir.glob("*.js")):
        try:
            text = f.read_text(encoding="utf-8", errors="ignore")
        except Exception:
            continue
        for m in re.finditer(r"""createTable\(\s*['"]([^'"]+)['"]""", text):
            names.add(m.group(1))
    return sorted(names)


def get_repo_overview(ctx: RepoContext) -> str:
    root = Path(ctx.repoPath)
    return "\n".join(
        [
            f"仓库路径：{root}",
            f"当前分支：{ctx.branch}",
            "--- 目录树（最多 2 层）---",
            "\n".join(_list_dir(root, "", 2)[:220]),
            "--- 既有数据库表名（来自 migrations 的 createTable，新建/修改 migration 必须照抄，注意大小写与单复数）---",
            ", ".join(_existing_db_tables(root)) or "(未发现)",
            "--- 后端 migration 文件 ---",
            "\n".join(_list_files(root, "backend/migrations")),
            "--- 前端入口文件 ---",
            f"--- frontend/src/App.jsx ---\n{_safe_read(root, 'frontend/src/App.jsx')}",
            f"--- frontend/src/main.jsx ---\n{_safe_read(root, 'frontend/src/main.jsx')}",
            "--- 前端页面文件 ---",
            "\n".join(_list_files(root, "frontend/src/routes")),
            "\n".join(_list_files(root, "frontend/src/routes/Profile")),
            "--- 后端路由注册 ---",
            _safe_read(root, "backend/index.js"),
            "--- 后端路由文件 ---",
            "\n".join(_list_files(root, "backend/routes")),
            "--- 后端模型文件 ---",
            "\n".join(_list_files(root, "backend/models")),
        ]
    )


def _clean_json(text: str) -> str:
    s = text.strip()
    if s.startswith("```"):
        s = s.split("\n", 1)[1] if "\n" in s else s
        if s.endswith("```"):
            s = s[:-3]
    return s.strip()


def steps_from_file_dicts(files: object) -> list[FileStep]:
    """校验并转换 LLM 返回的 files 数组为 FileStep 列表；超过保险上限时显式报错（不静默截断）。"""
    if not isinstance(files, list):
        raise ValueError("plan 没有返回 files 数组")
    if len(files) > MAX_GENERIC_FILES:
        raise ValueError(
            f"plan 规划了 {len(files)} 个文件，超过上限 {MAX_GENERIC_FILES}；"
            f"需求可能过大，建议拆分后分多次执行，或调高 MAX_GENERIC_FILES。"
        )
    steps: list[FileStep] = []
    for i, f in enumerate(files):
        if not isinstance(f, dict):
            raise ValueError(f"plan.files[{i}] 不是对象")
        path = str(f.get("path", "")).strip()
        mode = f.get("mode")
        instruction = str(f.get("instruction", "")).strip()
        if not path:
            raise ValueError(f"plan.files[{i}] 缺少 path")
        if mode not in {"modify", "create"}:
            raise ValueError(f"plan.files[{i}] mode 必须是 modify 或 create")
        if not instruction:
            raise ValueError(f"plan.files[{i}] 缺少 instruction")
        steps.append(FileStep(path=path, mode=mode, instruction=instruction))
    return steps


# 仅作为 locate/generate 的执行载体；规划与澄清统一走 plan_loop。
generic_skill = Skill(name="generic", description="通用推理执行载体（规划与澄清走 plan_loop）")
